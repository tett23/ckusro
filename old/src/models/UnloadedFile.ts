import { PromisifiedFS } from '../core/types';
import { LoaderContext } from './loaderContext';
import { FileModeFile, FileModes } from './StatType';

export type UnloadedFile = {
  loaderContext: LoaderContext;
  absolutePath: string;
  mode: FileModes;
};

export async function loadContent(
  readFile: PromisifiedFS['readFile'],
  unloadedFile: UnloadedFile,
): Promise<Buffer | string | null | Error> {
  if (unloadedFile.mode !== FileModeFile) {
    return null;
  }

  return readFile(unloadedFile.absolutePath).catch((err: Error) => err);
}
