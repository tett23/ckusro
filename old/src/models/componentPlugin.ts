import { isNonNullObject } from '../core/utils/types';
import { CkusroFile } from './CkusroFile';

export type Props = {
  currentFileId: string;
  files: CkusroFile;
};

export type ComponentPlugin = {
  name: string;
  plugin: (props: any) => JSX.Element;
};

export function isComponentPlugins(obj: any): obj is ComponentPlugin[] {
  if (!Array.isArray(obj)) {
    return false;
  }

  return obj.every(isComponentPlugin);
}

export function isComponentPlugin(obj: any): obj is ComponentPlugin {
  if (!isNonNullObject(obj)) {
    return false;
  }

  if (typeof obj.name !== 'string') {
    return false;
  }

  if (typeof obj.plugin !== 'function') {
    return false;
  }
  if (obj.plugin.length !== 1) {
    return false;
  }

  return true;
}
