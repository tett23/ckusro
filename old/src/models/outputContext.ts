import { join } from 'path';
import { isNonNullObject } from '../core/utils/types';
import { CkusroConfig } from './ckusroConfig';
import { LoaderContext } from './loaderContext';

export type OutputContext = {
  name: string;
  path: string;
};

export function newOutputContext(
  config: CkusroConfig,
  loaderContext: LoaderContext,
) {
  return {
    name: loaderContext.name,
    path: join(config.outputDirectory, loaderContext.name),
  };
}

export function isOutputContext(value: unknown): value is OutputContext {
  if (!isNonNullObject(value)) {
    return false;
  }

  if (typeof (value as OutputContext).name !== 'string') {
    return false;
  }

  if (typeof (value as OutputContext).path !== 'string') {
    return false;
  }

  return true;
}
