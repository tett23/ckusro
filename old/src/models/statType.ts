import { S_IFMT } from 'constants';

export const StatTypeBlockDevice: 'StatType:BlockDevice' =
  'StatType:BlockDevice';
export const StatTypeCharacterDevice: 'StatType:CharacterDevice' =
  'StatType:CharacterDevice';
export const StatTypeDirectory: 'StatType:Directory' = 'StatType:Directory';
export const StatTypeFIFO: 'StatType:FIFO' = 'StatType:FIFO';
export const StatTypeFile: 'StatType:File' = 'StatType:File';
export const StatTypeSocket: 'StatType:Socket' = 'StatType:Socket';
export const StatTypeSymbolicLink: 'StatType:SymbolicLink' =
  'StatType:SymbolicLink';
export type StatTypes =
  | typeof StatTypeBlockDevice
  | typeof StatTypeCharacterDevice
  | typeof StatTypeDirectory
  | typeof StatTypeFIFO
  | typeof StatTypeFile
  | typeof StatTypeSocket
  | typeof StatTypeSymbolicLink;

export const FileModeBlockDevice: 0o060000 = 0o060000; // S_IFBLK
export const FileModeCharacterDevice: 0o020000 = 0o020000; // S_IFCHR
export const FileModeDirectory: 0o040000 = 0o040000; // S_IFDIR
export const FileModeFIFO: 0o010000 = 0o010000; // S_IFIFO
export const FileModeFile: 0o100000 = 0o100000; // S_IFREG
export const FileModeSocket: 0o140000 = 0o140000; // S_IFSOCK
export const FileModeSymbolicLink: 0o120000 = 0o120000; // S_IFLNK

export type FileModes =
  | typeof FileModeBlockDevice
  | typeof FileModeCharacterDevice
  | typeof FileModeDirectory
  | typeof FileModeFIFO
  | typeof FileModeFile
  | typeof FileModeSocket
  | typeof FileModeSymbolicLink;

export function statType(mode: number): StatTypes {
  switch (
    S_IFMT & mode // tslint:disable-line no-bitwise
  ) {
    case FileModeBlockDevice:
      return StatTypeBlockDevice;
    case FileModeCharacterDevice:
      return StatTypeCharacterDevice;
    case FileModeDirectory:
      return StatTypeDirectory;
    case FileModeFIFO:
      return StatTypeFIFO;
    case FileModeFile:
      return StatTypeFile;
    case FileModeSocket:
      return StatTypeSocket;
    case FileModeSymbolicLink:
      return StatTypeSymbolicLink;
    default:
      throw new Error(`invalid stat mode ${mode}`);
  }
}
