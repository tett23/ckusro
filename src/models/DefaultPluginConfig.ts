import merge from 'lodash.merge';

export type DefaultPluginsConfig = {
  parsers: {
    enableWikiLink: boolean;
  };
  components: {
    enableWikiLink: boolean;
  };
};

export function defaultPluginsConfig(
  overrides: DeepPartial<DefaultPluginsConfig> = {},
): DefaultPluginsConfig {
  const data = {
    parsers: {
      enableWikiLink: true,
    },
    components: {
      enableWikiLink: true,
    },
  };

  return merge(data, overrides);
}
