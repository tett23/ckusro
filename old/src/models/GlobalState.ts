import { dirname, join } from 'path';
import LoaderInfoBuilder from '../core/LoaderInfoBuilder';
import NodeFS from '../core/LoaderInfoBuilder/NodeFS';
import { FS } from '../core/types';
import { separateErrors } from '../core/utils/errors';
import { isErrors, isNonNullObject } from '../core/utils/types';
import { CkusroConfig } from './ckusroConfig';
import { FileBuffersState } from './FileBuffersState';
import { LoaderContext, newLoaderContexts } from './loaderContext';
import { LocalLoaderContextType } from './loaderContext/LocalLoaderContext';
import { isNamespace, Namespace } from './Namespace';
import { newOutputContext } from './OutputContext';
import { isPlugins, Plugins } from './plugins';

export type GlobalState = {
  readonly namespaces: Namespace[];
  readonly plugins: Plugins;
};

export function isGlobalState(value: unknown): value is GlobalState {
  if (!isNonNullObject(value)) {
    return false;
  }

  if (!Array.isArray((value as GlobalState).namespaces)) {
    return false;
  }

  if (!(value as GlobalState).namespaces.every(isNamespace)) {
    return false;
  }

  if (!isPlugins((value as GlobalState).plugins)) {
    return false;
  }

  return true;
}

export function newGlobalState(conf: CkusroConfig): GlobalState {
  const namespaces = newLoaderContexts(
    conf.targetDirectories,
    conf.loaderConfig,
  ).map(
    (loaderContext: LoaderContext): Namespace => {
      return {
        name: loaderContext.name,
        loaderContext,
        outputContext: newOutputContext(conf, loaderContext),
      };
    },
  );

  return {
    namespaces,
    plugins: conf.plugins,
  };
}

function getFs(context: LoaderContext): FS {
  switch (context.type) {
    case LocalLoaderContextType:
      return NodeFS;
    default:
      throw new Error();
  }
}

export async function reloadFiles(
  globalState: GlobalState,
): Promise<FileBuffersState | Error[]> {
  const { plugins } = globalState;
  const ps = globalState.namespaces.map((ns) =>
    // FIXME
    LoaderInfoBuilder(getFs(ns.loaderContext), ns, plugins),
  );
  const results: Array<FileBuffersState | Error> = (await Promise.all(
    ps,
  )).flat() as any[];
  const [fsbs, errors] = separateErrors(results);
  if (isErrors(errors)) {
    return errors;
  }

  return fsbs.reduce(
    (acc, fbs) => {
      acc.fileBuffers.push(...fbs.fileBuffers);
      acc.dependencyTable = { ...acc.dependencyTable, ...fbs.dependencyTable };
      acc.dependencyTable = {
        ...acc.invertedDependencyTable,
        ...fbs.invertedDependencyTable,
      };

      return acc;
    },
    {
      fileBuffers: [],
      dependencyTable: {},
      invertedDependencyTable: {},
    } as FileBuffersState,
  );
}

export function outputDirectory(globalState: GlobalState): string {
  return dirname(globalState.namespaces[0].outputContext.path);
}

export function assetsDirectory(globalState: GlobalState): string {
  return join(outputDirectory(globalState), 'assets');
}
