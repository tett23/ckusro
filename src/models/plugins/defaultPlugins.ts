import { DefaultPluginsConfig } from '../../models/pluginConfig';
import WikiLink from '../../plugins/ckusro-plugin-component-WikiLink/WikiLink';
import wikiLink from '../../plugins/ckusro-plugin-parser-WikiLink';
import { Plugins } from './index';

export default function defaultPlugins(config: DefaultPluginsConfig): Plugins {
  const ret: Plugins = {
    parsers: [],
    components: [],
  };

  if (config.parsers.enableWikiLink) {
    ret.parsers.push({
      name: wikiLink.name,
      plugin: wikiLink,
    });
  }

  if (config.components.enableWikiLink) {
    ret.components.push({
      name: WikiLink.name,
      plugin: WikiLink,
    });
  }

  return ret;
}
