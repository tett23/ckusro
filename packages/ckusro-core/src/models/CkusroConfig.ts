import { Plugins } from './plugins';

export type CkusroConfig = {
  base: string;
  coreId: string;
  corsProxy: string | null;
  colorScheme: ColorScheme;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  plugins: Plugins<any, any>;
  authentication: {
    github: string | null;
  };
};

export type ColorScheme = {
  main: number;
  accent: number;
  background: number;
  base: number;
  text: number;
};

export type ColorSchemeText = { [K in keyof ColorScheme]: string };

export function convertColorScheme(colorScheme: ColorSchemeText): ColorScheme {
  return (Object.entries(colorScheme) as Array<
    [keyof ColorScheme, string]
  >).reduce(
    (acc, [key, value]) => {
      acc[key] = parseInt(value.replace(/^#/, ''), 16);

      return acc;
    },
    // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
    {} as ColorScheme,
  );
}
