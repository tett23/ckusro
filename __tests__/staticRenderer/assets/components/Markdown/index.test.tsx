import { shallow } from 'enzyme';
import React from 'react';
import { defaultPluginsConfig } from '../../../../../src/models/DefaultPluginConfig';
import defaultPlugins from '../../../../../src/models/plugins/defaultPlugins';
import {
  buildJSX,
  Markdown,
} from '../../../../../src/staticRenderer/assets/components/Markdown';
import { buildFile } from '../../../../__fixtures__';

describe(Markdown, () => {
  const plugins = defaultPlugins(defaultPluginsConfig());

  it('renders correctly', () => {
    const files = [buildFile({ id: 'test1', content: '[[foo]]' })];
    const actual = shallow(
      <Markdown plugins={plugins} currentFileId={files[0].id} files={files} />,
    );

    expect(actual).toMatchSnapshot();
  });

  it('throw Error when file does not exist', () => {
    const actual = () =>
      shallow(
        <Markdown
          plugins={plugins}
          currentFileId="does_not_exist"
          files={[]}
        />,
      );

    expect(actual).toThrowError();
  });
});

describe(buildJSX, () => {
  const plugins = defaultPlugins(defaultPluginsConfig());

  it('renders correctly', () => {
    const actual = buildJSX(plugins, '[[foo]]');

    expect(actual).toMatchSnapshot();
  });
});
