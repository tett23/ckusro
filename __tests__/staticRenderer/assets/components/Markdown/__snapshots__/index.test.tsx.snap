// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown renders correctly 1`] = `ShallowWrapper {}`;

exports[`buildJSX renders correctly 1`] = `
<p>
  <span
    data-reactroot=""
  >
    foo
  </span>
</p>
`;
